libflux: support flux_get_process_scope()

chu11 opened the pull request:
Per #3817.  This is super WIP with no testing yet, but wanted to post to get initial high level feedback.

- basic logic within function for the 5 categories discussed in #3817
- should be 4 categories instead of 5?  some discussion for this in #3817 too
- API style / enum names
- location of the function, I put it in new files `libflux/util.[ch]` ... but could go in `libjob` instead? but that's not the super right place either ... 

TODO
- the command
- tons of tests
- do we need a nice python wrapper?
- document it / add nice comments & examples

grondo commented on src/common/libflux/util.h:
```diff
@@ -0,0 +1,39 @@
+extern "C" {
+#endif
+
+typedef enum flux_process_scope {
+    FLUX_PROCESS_SCOPE_NONE = 0,
+    FLUX_PROCESS_SCOPE_SYSTEM_INSTANCE = 1,
+    FLUX_PROCESS_SCOPE_USER_INSTANCE_INITIAL_PROGRAM = 2,
+    FLUX_PROCESS_SCOPE_FLUX_JOB_INSTANCE_INITIAL_PROGRAM = 3,
+    FLUX_PROCESS_SCOPE_FLUX_JOB_INSTANCE_JOB = 4,
+    FLUX_PROCESS_SCOPE_UNKNOWN = 99,
+} flux_process_scope_t;
```
Not a full review, just wanted to get this comment in for consideration:

I would start with a pared down list of scopes with less egregiously long names.  :wink: I _think_ this is all our use cases need at this point? Callers that need to differentiate can do so via other attributes, etc.

```suggestion
typedef enum flux_process_scope {
    FLUX_PROCESS_SCOPE_NONE = 0,
    FLUX_PROCESS_SCOPE_SYSTEM = 1,
    FLUX_PROCESS_SCOPE_INITIAL_PROGRAM = 2,
    FLUX_PROCESS_SCOPE_JOB = 3,
    FLUX_PROCESS_SCOPE_UNKNOWN = 99,
} flux_process_scope_t;
```

Also, I didn't look at the code yet, is there really a need for `FLUX_PROCESS_SCOPE_UNKNOWN`? Shouldn't the function just return an error if it cannot determine the scope?

chu11 replied:
> Shouldn't the function just return an error if it cannot determine the scope?

Because we don't cover every permutation (per my comment in #3817), I wanted to have some flexibility to allow for future additions, thus the UNKNOWN and why I set that to 99.  We could just return an error instead.  Hmmmm EPERM?  Will have to think about it.

grondo replied:
Aren't the extra permutations just special cases of the already covered scopes as described above?

chu11 replied:
yeah, that's true.  given the more limited set, the potential for future extensions is less likely.

chu11 commented:
re-pushed with a hopefully reviewable version of this

- there's `flux_get_process_scope()`.  Again, I stuck it in a new `libflux/util.[ch]`, which I dunno if that's the best place.  In `libjob` would maybe be reasonable too, I waffled back and forth on it.  Could be in `scope.[ch]` too as `util` might be too generic?
- added some tests
- added a `flux scope` command
- added documentation

left to debate, do we need a python utility function for this?  in `util.py` seems wrong.  Maybe that'll determine where it goes in the C code too.

chu11 commented:
per comment in #3817 from @garlick, closing
